controllers/volumereplicationgroup_controller.go

```diff
@@ -197,77 +207,284 @@ func (v *VolumeReplicationGroupReconciler) Reconcile(ctx context.Context, req ct
 		return ctrl.Result{}, fmt.Errorf("failed to reconcile VolumeReplicationGroup %w", err)
 	}
 
-	log.Info("Processing VolumeReplicationGroup ", volRepGroup.Spec, " in ns: ", req.NamespacedName)
+	v.ctx = ctx
 
-	pvcList, err := v.HandlePersistentVolumeClaims(ctx, volRepGroup)
-	if err != nil {
+	if err := v.updatePVCList(); err != nil {
 		log.Error(err, "Handling of Persistent Volume Claims of application failed")
 
 		return ctrl.Result{}, err
 	}
 
+	log.Info("Processing VolumeReplicationGroup ", v.instance.Spec, " in ns: ", req.NamespacedName)
+
+	switch {
+	case !v.instance.GetDeletionTimestamp().IsZero():
+		return v.finalizeVRG()
+	case v.instance.Spec.ReplicationState == volrep.ReplicationPrimary:
+		return v.processAsPrimary()
+	case v.instance.Spec.ReplicationState == volrep.ReplicationSecondary:
+		return v.processAsSecondary()
+	}
+
+	// No requeue as there is no reconcile till user changes desired spec to a valid value
+	return ctrl.Result{Requeue: false},
+		fmt.Errorf("invalid or unknown replication state detected, deleted %v, desired replicationState %v",
+			v.instance.GetDeletionTimestamp().IsZero(),
+			v.instance.Spec.ReplicationState)
+}
+
+func (v *VRGInstance) finalizeVRG() (ctrl.Result, error) {
+	if !containsString(v.instance.ObjectMeta.Finalizers, vrgFinalizerName) {
+		return ctrl.Result{}, nil
+	}
+
+	if err := v.finalizeChildren(); err != nil {
+		log.Errorf("failed to finalize child resources: %s/%s", v.instance.Name, v.instance.Namespace)
+
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if err := v.removeFinalizer(vrgFinalizerName); err != nil {
+		log.Errorf("failed to remove the finalizer from vrg: %s/%s", v.instance.Name, v.instance.Namespace)
+
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// stop reconciliation as the item is being deleted
+	return ctrl.Result{}, nil
+}
+
+// processAsPrimary reconciles the current instance of VRG as primary
+/* Primary:
+  - For each PVC
+	- If not bound, continue but requeue
+    * Handle PVC being Deleted?
+	- Check `vr-replicated` finalizer, if found skip to checking status of PVC
+	- Add `vr-protection` finalizer to PVC (to delete/update VR when PVC is deleted)
+    - [Find|Create] VR for PVC as Primary
+	- Backup PV bound to PVC
+	- Add `vr-replicated` finalizer to PVC
+  - For each PVC
+    - Check if VR status has reached Primary state
+		- If not, update status as progressing
+  - Report status as completed, if not progressing for any PVC
+* For PVCs being deleted with VRG as primary, we need to handle garbage collection of PVs
```

Comment:
(nit) Again a minor clarification. I assume "we need to handle garbage collection of PVs" as "we need to handle garbage collection of backedup PV metadata" ?

Reply:
Ah! no :)

The PVs are created with the reclaimPolicy "Retain" so that these are not deleted. We may end up modifying the retention policy as well when we create a VR->PVC->PV, so this really is changing the reclaim policy to "Delete" when we are unlinking a VR from a PVC.

Well, we also need to cleanup the PV in the S3 store :)